File: CS/WindowsApplication1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;

namespace WindowsApplication1
{
    public partial class Form1 : Form
    {
        private DataTable CreateTable(int RowCount)
        {
            DataTable tbl = new DataTable();
            tbl.Columns.Add("Name", typeof(string));
            tbl.Columns.Add("ID", typeof(int));
            tbl.Columns.Add("Number", typeof(int));
            tbl.Columns.Add("Date", typeof(DateTime));
            tbl.Columns.Add("ParentID", typeof(int));
            for (int i = 0; i < RowCount; i++)
                tbl.Rows.Add(new object[] { String.Format("Name{0}", i), i + 1, 3 - i, DateTime.Now.AddDays(i), i % 3 });
            return tbl;
        }

        HideCheckBoxHelper hideCheckBoxHelper;
        public Form1()
        {
            InitializeComponent();
            treeList1.DataSource = CreateTable(30);
            treeList1.ExpandAll();
            hideCheckBoxHelper = new HideCheckBoxHelper(treeList1);
        }

        private void spinEdit1_EditValueChanged(object sender, EventArgs e)
        {
            hideCheckBoxHelper.Level = Convert.ToInt16(spinEdit1.Value);
        }

        private void checkEdit1_CheckedChanged(object sender, EventArgs e)
        {
            hideCheckBoxHelper.NeedHide = checkEdit1.Checked;
        }
    }

    public class HideCheckBoxHelper
    {
        
        public HideCheckBoxHelper(TreeList treeList)
        {
            _TreeList = treeList;
            treeList.CustomDrawNodeCheckBox += treeList_CustomDrawNodeCheckBox;
            treeList.BeforeCheckNode += treeList_BeforeCheckNode;
        }

        private int _Level = -1;
        private TreeList _TreeList;
        public int Level
        {
            get { return _Level; }
            set { _Level = value; _TreeList.Refresh(); }
        }

        private bool _Hide = true;
        public bool NeedHide
        {
            get { return _Hide; }
            set { _Hide = value; _TreeList.Refresh(); }
        }

        private bool CanCheckNode(TreeListNode node)
        {
            if (Level == -1)
                return true;
            return node.Level == Level;
        }

        void treeList_BeforeCheckNode(object sender, CheckNodeEventArgs e)
        {
            e.CanCheck = CanCheckNode(e.Node);
        }

        void treeList_CustomDrawNodeCheckBox(object sender, CustomDrawNodeCheckBoxEventArgs e)
        {
            bool canCheckNode = CanCheckNode(e.Node);
            if (canCheckNode)
                return;
            e.ObjectArgs.State = DevExpress.Utils.Drawing.ObjectState.Disabled;
            e.Handled = NeedHide;
        }
    }
}

File: VB/WindowsApplication1/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes

Namespace WindowsApplication1

    Public Partial Class Form1
        Inherits Form

        Private Function CreateTable(ByVal RowCount As Integer) As DataTable
            Dim tbl As DataTable = New DataTable()
            tbl.Columns.Add("Name", GetType(String))
            tbl.Columns.Add("ID", GetType(Integer))
            tbl.Columns.Add("Number", GetType(Integer))
            tbl.Columns.Add("Date", GetType(Date))
            tbl.Columns.Add("ParentID", GetType(Integer))
            For i As Integer = 0 To RowCount - 1
                tbl.Rows.Add(New Object() {String.Format("Name{0}", i), i + 1, 3 - i, Date.Now.AddDays(i), i Mod 3})
            Next

            Return tbl
        End Function

        Private hideCheckBoxHelper As HideCheckBoxHelper

        Public Sub New()
            InitializeComponent()
            treeList1.DataSource = CreateTable(30)
            treeList1.ExpandAll()
            hideCheckBoxHelper = New HideCheckBoxHelper(treeList1)
        End Sub

        Private Sub spinEdit1_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            hideCheckBoxHelper.Level = Convert.ToInt16(spinEdit1.Value)
        End Sub

        Private Sub checkEdit1_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            hideCheckBoxHelper.NeedHide = checkEdit1.Checked
        End Sub
    End Class

    Public Class HideCheckBoxHelper

        Public Sub New(ByVal treeList As TreeList)
            _TreeList = treeList
            AddHandler treeList.CustomDrawNodeCheckBox, AddressOf treeList_CustomDrawNodeCheckBox
            AddHandler treeList.BeforeCheckNode, AddressOf treeList_BeforeCheckNode
        End Sub

        Private _Level As Integer = -1

        Private _TreeList As TreeList

        Public Property Level As Integer
            Get
                Return _Level
            End Get

            Set(ByVal value As Integer)
                _Level = value
                _TreeList.Refresh()
            End Set
        End Property

        Private _Hide As Boolean = True

        Public Property NeedHide As Boolean
            Get
                Return _Hide
            End Get

            Set(ByVal value As Boolean)
                _Hide = value
                _TreeList.Refresh()
            End Set
        End Property

        Private Function CanCheckNode(ByVal node As TreeListNode) As Boolean
            If Level = -1 Then Return True
            Return node.Level = Level
        End Function

        Private Sub treeList_BeforeCheckNode(ByVal sender As Object, ByVal e As CheckNodeEventArgs)
            e.CanCheck = CanCheckNode(e.Node)
        End Sub

        Private Sub treeList_CustomDrawNodeCheckBox(ByVal sender As Object, ByVal e As CustomDrawNodeCheckBoxEventArgs)
            Dim canCheckNode As Boolean = Me.CanCheckNode(e.Node)
            If canCheckNode Then Return
            e.ObjectArgs.State = DevExpress.Utils.Drawing.ObjectState.Disabled
            e.Handled = NeedHide
        End Sub
    End Class
End Namespace
